core/node/libp2p/discovery.go

```diff
@@ -0,0 +1,52 @@
+package libp2p
+
+import (
+	"context"
+	"time"
+
+	"github.com/ipfs/go-ipfs-config"
+	"github.com/ipfs/go-ipfs/core/node/helpers"
+	"github.com/libp2p/go-libp2p-host"
+	"github.com/libp2p/go-libp2p-peerstore"
+	"github.com/libp2p/go-libp2p/p2p/discovery"
+	"go.uber.org/fx"
+)
+
+const discoveryConnTimeout = time.Second * 30
+
+type discoveryHandler struct {
+	ctx  context.Context
+	host host.Host
+}
+
+func (dh *discoveryHandler) HandlePeerFound(p peerstore.PeerInfo) {
+	log.Warning("trying peer info: ", p)
+	ctx, cancel := context.WithTimeout(dh.ctx, discoveryConnTimeout)
+	defer cancel()
+	if err := dh.host.Connect(ctx, p); err != nil {
+		log.Warning("Failed to connect to peer found by discovery: ", err)
+	}
+}
+
+func DiscoveryHandler(mctx helpers.MetricsCtx, lc fx.Lifecycle, host host.Host) *discoveryHandler {
```

Comment:
exported func DiscoveryHandler returns unexported type *libp2p.discoveryHandler, which can be annoying to use